Mathlib/NumberTheory/ModularForms/QExpansion.lean

```diff
@@ -89,108 +102,239 @@ theorem differentiableAt_comp_ofComplex [ModularFormClass F Γ k]
+  -- We use a circle integral in the `q`-domain of radius `R = exp (-2 * π * t / N)`.
+  let R := Real.exp (-2 * π * t / h)
+  have hR0 : 0 < R := Real.exp_pos _
+  have hR1 : R < 1 := Real.exp_lt_one_iff.mpr <| by
+    simp only [neg_mul, neg_div, neg_lt_zero]
+    exact div_pos (by positivity) hh
+  -- First apply `qExpansion_coeff_eq_circleIntegral` and rescale from `0 .. 2 * π` to `0 .. N`.
+  rw [qExpansion_coeff_eq_circleIntegral f hh hΓ n hR0 hR1, circleIntegral,
+    show 2 * π = h * (2 * π / h) by field_simp [NeZero.ne]]
+  conv => enter [1, 2, 2]; rw [show 0 = 0 * (2 * π / h) by simp]
+  simp_rw [← intervalIntegral.smul_integral_comp_mul_right, real_smul, ← mul_assoc,
+    ← intervalIntegral.integral_const_mul]
+  -- Compare the integrands
+  congr 1 with u
+  let τ : ℍ := ⟨u + t * I, by simpa using ht⟩
+  have : circleMap 0 R (u * (2 * π / h)) = 𝕢 h τ := by
+    simp only [circleMap, ofReal_exp, ← exp_add, zero_add, τ, UpperHalfPlane.coe_mk_subtype, R]
+    congr 1
+    push_cast
+    have := I_sq
+    grind
+  -- now just complex exponential arithmetic to finish
+  simp_rw [deriv_circleMap, this, show u + t * I = τ by rfl, show ⟨↑τ, τ.2⟩ = τ by rfl,
+    eq_cuspFunction f _ hΓ hh.ne', smul_eq_mul, pow_succ, push_cast]
+  field_simp [(show 𝕢 h τ ≠ 0 from Complex.exp_ne_zero _), Real.pi_ne_zero, NeZero.ne]
+
+lemma isCusp_of_mem_strictPeriods (hh : 0 < h) (hΓ : h ∈ Γ.strictPeriods)
+    [DiscreteTopology Γ] [Γ.HasDetPlusMinusOne] :
+    IsCusp OnePoint.infty Γ := by
+  rw [Subgroup.strictPeriods_eq_zmultiples_strictWidthInfty] at hΓ
+  refine Γ.strictWidthInfty_pos_iff.mp <| Γ.strictWidthInfty_nonneg.lt_of_ne' fun h0 ↦ hh.ne' ?_
+  simpa only [h0, AddSubgroup.zmultiples_zero_eq_bot, AddSubgroup.mem_bot] using hΓ
+
+theorem exp_decay_sub_atImInfty [ModularFormClass F Γ k] [Γ.HasDetPlusMinusOne]
+    [DiscreteTopology Γ] (hh : 0 < h) (hΓ : h ∈ Γ.strictPeriods) :
+    (fun τ ↦ f τ - valueAtInfty f) =O[atImInfty] fun τ ↦ Real.exp (-2 * π * τ.im / h) := by
+  have hi : IsCusp OnePoint.infty Γ := isCusp_of_mem_strictPeriods hh hΓ
+  convert ((periodic_comp_ofComplex f hΓ).exp_decay_sub_of_bounded_at_inf hh
+    (eventually_of_mem (preimage_mem_comap (Ioi_mem_atTop 0))
+        fun _ ↦ differentiableAt_comp_ofComplex f)
+    (bounded_at_infty_comp_ofComplex f hi)).comp_tendsto tendsto_coe_atImInfty
+  simp only [comp_apply, ofComplex_apply, ← cuspFunction_apply_zero f hh hΓ, cuspFunction]
+
+/-- Version of `exp_decay_sub_atImInfty` stating a less precise result but easier to apply in
+practice (not specifying the growth rate precisely). Note that the `Fact` hypothesis is
+automatically synthesized for arithmetic subgroups. -/
+theorem exp_decay_sub_atImInfty' [ModularFormClass F Γ k] [Γ.HasDetPlusMinusOne]
+    [DiscreteTopology Γ] [Fact (IsCusp OnePoint.infty Γ)] :
+    ∃ h > 0, (fun τ ↦ f τ - valueAtInfty f) =O[atImInfty] (fun τ ↦ Real.exp (-h * τ.im)) := by
```

Comment:
Maybe using `h` here is not optimal, as it is used before to denote a period.
(Note that in the preceding result, it is actually `2 * π / h`...)